setup_unit_tests.py

```diff
@@ -0,0 +1,488 @@
+            # else check if there is extra metadata to read from
+            if get_not_none(metadata, "md5sum") is not None:
+                return {'md5sum': metadata.get("md5sum")}
+            if get_not_none(metadata, "regex") is not None:
+                return {'regex': metadata.get("regex")}
+
+    if isinstance(output_type, WDL.Type.StructInstance):
+        converted_output = dict()
+        for i, (output_key, output_value) in enumerate(output_values.items()):
+            try:
+                output_value_type = output_type.members[output_key]
+            except KeyError:
+                # coercion from map to struct can cause some weird type behavior
+                # since dictionaries past python 3.6 are ordered, find the corresponding type from the current output's position
+                output_value_type = list(output_type.members.values())[i]
+            converted_output[output_key] = convert_typed_output_values(output_value, output_value_type, data_dir, get_not_none(metadata, output_value))
+        return converted_output
+    if isinstance(output_type, WDL.Type.Map):
+        converted_output = dict()
+        for output_key, output_value in output_values.items():
+            new_output_key = convert_typed_output_values(output_key, output_type.item_type[0], data_dir, get_not_none(metadata, output_key))
+            new_output_value = convert_typed_output_values(output_value, output_type.item_type[1], data_dir, get_not_none(metadata, output_value))
+            converted_output[new_output_key] = new_output_value
+    if isinstance(output_type, WDL.Type.Pair):
+        converted_output = dict()
+        # key should be left or right
+        converted_output["left"] = convert_typed_output_values(output_values["left"], output_type.left_type, data_dir, get_not_none(metadata, output_values["left"]))
+        converted_output["right"] - convert_typed_output_values(output_values["right"], output_type.right_type, data_dir, get_not_none(metadata, output_values["right"]))
+        return converted_output
+    if isinstance(output_type, WDL.Type.Array):
+        converted_output = list()
+        for i, output in enumerate(output_values):
+            converted_output.append(convert_typed_output_values(output, output_type.item_type, data_dir, index_not_none(metadata, i)))
+        return converted_output
+    # else this is a primitive type
+    return output_values
+
+
+def convert_typed_output_values_from_string(output_values: Union[None, str, Dict[str, Any], List[Any]], output_type: Union[str, Dict[str, Any]],
+                                            data_dir: Optional[Path], metadata: Optional[Dict[str, Any]]):
+    output_type_wdl = convert_type(output_type)
+    return convert_typed_output_values(output_values, output_type_wdl, data_dir, metadata)
+
+
+def recursive_json_apply(json_obj: Union[Dict[Any, Any], List[Any], bool, int, str, float, None], func: Callable[[Any], Any]) \
```

Comment:
Maybe this type should be stored into a variable named something to indicate we mean any parsed JSON thing?

Reply:
Also I think I know what this is meant to do but it could use a docstring.

Reply:
Added a docstring